### DISCUSSION TITLE
Patrol routes

### DISCUSSION THREAD

**Discussion by Roms1383** [Halen84/ScriptHookRDR2DotNet-V2#10] [Q&A] [answered]
Since we already discussed it in #9, I'm linking this discussion to it as a reminder (especially when the issue will be closed).

---

Here was my question:

*I'd like more informations about patrol routes:
as far as I understand looking at the definitions, a patrol route can be created, added nodes and links, opened, closed an deleted.
Something that I quite don't get though, is that I would expect `TASK::CREATE_PATROL_ROUTE` to return a handle, since a `patrolRoute` parameter is expected in the other methods. Any clue on how to work this out ?*

And here was @Halen84's reply (thanks so much for it 🙏 ❤️ ):

*As for `TASK::CREATE_PATROL_ROUTE`, it doesn't return anything because the creation of nodes and links is handled in the engine for the script threads, so the native probably also handles the creation of it in the engine. So basically, its all handled in the engine.*

_Looking at the decompiled scripts, it looks like you first have to follow these steps in order:_

- _Open a patrol route with `TASK::OPEN_PATROL_ROUTE()`_
- _Create your nodes with `TASK::ADD_PATROL_ROUTE_NODE()`_
- _Link the created nodes with `TASK::ADD_PATROL_ROUTE_LINK()`_
- _`TASK::CLOSE_PATROL_ROUTE()` when you have finished creating and linking nodes_
- _`TASK::CREATE_PATROL_ROUTE()` when all of the above has been finished_
- _`TASK::DELETE_PATROL_ROUTE()` when the patrol route is no longer needed. This will be needed so the engine can free up any allocated resources created by these natives._

_Very similar process to task sequences for example._

**Comment by Roms1383** [comment]
A quick follow-up on `patrol routes` since I gave a (quick) try yesterday : I couldn't get the methods call right e.g. for `TASK::ADD_PATROL_ROUTE_NODE` and the likes. The game crashed whenever I attempted to call methods to build a custom patrol route.

Looking at its definition :
- with `dnSpy` :
   ```csharp
    void ADD_PATROL_ROUTE_NODE(ulong p0, ulong p1, ulong p2, ulong p3, ulong p4, ulong p5, ulong p6, ulong p7, ulong p8, ulong p9)
    ```
- on [alloc8or](https://alloc8or.re/rdr3/nativedb/?n=0x8EDF950167586B7C) :
   ```csharp
    void ADD_PATROL_ROUTE_NODE(Any p0, Any p1, Any p2, Any p3, Any p4, Any p5, Any p6, Any p7, Any p8, Any p9)
    ```

So I went to look at potential [definition on FiveM](https://docs.fivem.net/natives/?_0x8EDF950167586B7C), since both GTA and RDR's game engines are pretty similar, which yields a bit more informations though:
```csharp
void AddPatrolRouteNode(int p0, string p1, float x1, float y1, float z1, float x2, float y2, float z2, int p8);
```
but more importantly there's an associated comment:
```
TASK::ADD_PATROL_ROUTE_NODE(2, "WORLD_HUMAN_GUARD_STAND", -193.4915, -2378.864990234375, 10.9719, -193.4915, -2378.864990234375, 10.9719, 3000);
p0 is between 0 and 4 in the scripts.
p1 is "WORLD_HUMAN_GUARD_STAND" or "StandGuard".
p2, p3 and p4 is only one parameter sometimes in the scripts. Most likely a Vector3 hence p2, p3 and p4 are coordinates.
Examples:
TASK::ADD_PATROL_ROUTE_NODE(1, "WORLD_HUMAN_GUARD_STAND", l_739[7/*3*/], 0.0, 0.0, 0.0, 0);
TASK::ADD_PATROL_ROUTE_NODE(1, "WORLD_HUMAN_GUARD_STAND", l_B0[17/*44*/]._f3, l_B0[17/*44*/]._f3, 2000);
p5, p6 and p7 are for example set to: 1599.0406494140625, 2713.392578125, 44.4309.
p8 is an int, often random set to for example: MISC::GET_RANDOM_INT_IN_RANGE(5000, 10000).
```

An example of the code I tried so far would be:
```csharp
public void Patrol()
{
        if (this.ped != null && !this.ped.IsDeadOrDying)
        {
            Vector3 pos = ENTITY.GET_ENTITY_COORDS(this.ped, true, true);
            Vector3 pos2 = new Vector3(pos.X + 10, pos.Y + 10, pos.Z);
            Vector3 pos3 = new Vector3(pos.X + 40, pos.Y + 40, pos.Z);
            string id = "patrol_area_custom";
            ulong hash = MISC.GET_HASH_KEY(id);
            TASK.OPEN_PATROL_ROUTE(id);
            TASK.ADD_PATROL_ROUTE_NODE(0, hash, (ulong)pos.X, (ulong)pos.Y, (ulong)pos.Z, 1, 1, 1, 1, (ulong) MISC.GET_RANDOM_INT_IN_RANGE(5000, 10000)); ;
            TASK.ADD_PATROL_ROUTE_NODE(1, hash, (ulong)pos.X, (ulong)pos.Y, (ulong)pos.Z, 2, 2, 2, 2, (ulong) MISC.GET_RANDOM_INT_IN_RANGE(5000, 10000));
            TASK.ADD_PATROL_ROUTE_NODE(2, hash, (ulong)pos.X, (ulong)pos.Y, (ulong)pos.Z, 3, 3, 3, 3, (ulong) MISC.GET_RANDOM_INT_IN_RANGE(5000, 10000));
            TASK.ADD_PATROL_ROUTE_LINK(0, 1);
            TASK.ADD_PATROL_ROUTE_LINK(1, 2);
            TASK.ADD_PATROL_ROUTE_LINK(2, 0);
            TASK.CLOSE_PATROL_ROUTE();
            TASK.CREATE_PATROL_ROUTE();
            TASK.TASK_PATROL(this.ped, id, 10, true, true);
       }
}
```

**Reply by Halen84** [reply]
On alloc8or's db, `Any`  is a type alias for `uint64_t` (`unsigned long long`) in C++, or a `ulong` (`System.UInt64`) in C#. So its the exact same thing. `Any` or `ulong` means the actual type of the param isn't documented.

Instead of doing explicit casts on the parameters like `float`s (you will lose data doing a cast), you can instead use the `Function.Call()` function to call it with correct parameters without losing data.

Example:
```cs
// TASK::ADD_PATROL_ROUTE_NODE
Function.Call(0x8EDF950167586B7C, 0, "patrol_area_custom", pos.X, pos.Y, pos.Z, 1, 1, 1, 1, MISC.GET_RANDOM_INT_IN_RANGE(5000, 10000));
```

**Reply by Roms1383** [reply]
> On alloc8or's db, Any is a type alias for uint64_t (unsigned long long) in C++, or a ulong (System.UInt64) in C#. So its the exact same thing. Any or ulong means the actual type of the param isn't documented.

That's what I kinda guessed, but had no idea if they were pointers or not, thanks for the clarification.

I'm currently giving `Function.Call` a try and also modified the calls like so:
```csharp
public void Patrol()
{
    RDR2.UI.Screen.PrintSubtitle("patrol ?");
    if (this.ped != null && !this.ped.IsDeadOrDying)
    {
        RDR2.UI.Screen.PrintSubtitle("patrol !");
        Vector3 pos = ENTITY.GET_ENTITY_COORDS(this.ped, true, true);
        Vector3 pos2 = new Vector3(pos.X + 10, pos.Y + 10, pos.Z);
        Vector3 pos3 = new Vector3(pos.X + 40, pos.Y + 40, pos.Z);
        string id = "patrol_area_custom";
        ulong hash = MISC.GET_HASH_KEY(id);
        int rand = MISC.GET_RANDOM_INT_IN_RANGE(5000, 10000);
        TASK.OPEN_PATROL_ROUTE(id);
        // TASK.ADD_PATROL_ROUTE_NODE
        Function.Call(0x8EDF950167586B7C, 0, "patrol_area_custom", pos.X, pos.Y, pos.Z, pos.X, pos.Y, pos.Z, rand, true);
        Function.Call(0x8EDF950167586B7C, 1, "patrol_area_custom", pos2.X, pos2.Y, pos2.Z, pos2.X, pos2.Y, pos2.Z, rand, true);
        Function.Call(0x8EDF950167586B7C, 2, "patrol_area_custom", pos3.X, pos3.Y, pos3.Z, pos3.X, pos3.Y, pos3.Z, rand, true);
        TASK.ADD_PATROL_ROUTE_LINK(0, 1);
        TASK.ADD_PATROL_ROUTE_LINK(1, 2);
        TASK.ADD_PATROL_ROUTE_LINK(2, 0);
        TASK.CLOSE_PATROL_ROUTE();
        TASK.CREATE_PATROL_ROUTE();
        TASK.TASK_PATROL(this.ped, id, 10, true, true);
    }
}
```

I basically put the same `Vector3` twice since it's what it looks like in `FiveM` comment. Also switched the last param for a `bool` since FiveM signature only has 8 vs 9 params found in `RDR`, and from what I've seen, most of R* methods end with a boolean (yet I have no idea what it does yet).

Good news is that it doesn't crash the game anymore (it's entering the patrol creation, as asserted by the `patrol !` log), bad news is that the ped just stand still and doesn't start its patrol yet ^^'

**Reply by Halen84** [reply, marked as answer]
Looks like I messed up. Patrol route name needs to be prefixed with "miss_" for it be valid, and the string parameter is actually a scenario.

Correct C++ definitions:
```cpp
void TASK::ADD_PATROL_ROUTE_NODE(int nodeId, const char* scenarioName, float x, float y, float z, float lookPosX, float lookPosY, float lookPosZ, int duration, BOOL p9);
void TASK::ADD_PATROL_ROUTE_LINK(int node1, int node2);
```

**Reply by Roms1383** [reply]
Ok apparently the name just need to be prefixed with `miss_`, even `miss_patrol_area_custom` seems to do the trick.
What is somehow weird is that if the ped is "off path" (like not located on a node of the patrol route), he will try to join the closest node, but apparently then he goes only back and forth between 2 nodes.
On the other hand if I spawn him directly on a node, then it seems he's just standing still.

**Reply by Halen84** [reply]
That might be an issue with how the nodes are linked. I've never messed with patrol routes so I wouldn't know

**Reply by Roms1383** [reply]
I'll run a couple more experiments later this week and report here if ever ^^ thanks for your help I'm gonna mark it as resolved since there's enough to get it running already.
